_posts/resources/2023-05-08-morello-linux-morelloie.md

```diff
@@ -0,0 +1,1096 @@
+```
+
+The GCC toolchain doesn’t require you to specify a path to sysroot explicitly or use target triple, however you need to explicitly mention target architecture and ABI because the same toolchain may build binaries for AArch64 as well as for Morello.
+
+Whatever toolchain you’re using, you will end up with a binary that is what we call a purecap Morello application. We’ll discuss the details of this later, but right now it’s important that when we run it, we’ll be using Morello, which is exactly what we want. To verify this, we can use the `llvm-readelf` tool (or just `readelf` if you use GNU toolchain):
+
+```
+$ llvm-readelf -h hello
+...
+  Machine:                           AArch64
+  Entry point address:               0x210121
+...
+  Flags:                             0x10000, purecap
+...
+```
+
+Some output above is omitted. Here we may notice two things: the entry point address `0x210121` is odd (pun indented), and the flags field contains `0x10000` which means "purecap" (see [Morello ELF spec][ELF]). Each of these things tells us that this is a purecap binary.
+
+The `hello` binary you have now would run in the same way on any Morello platform, to run it on Morello IE, use:
+
+```
+$ morelloie -- ./hello
+Hello Morello
+```
+
+Double hyphen separates emulator’s parameters from the target application’s command line.
+
+Well done! We’ve run our first Morello app and it works! Surely, there are simpler ways to print "hello world". In fact, where is Morello in all this? Yes, we used a Morello toolchain to build the binary, but the code still looks the same as you would use for any other target platform. Well, that's the thing about Morello (and CHERI in general): normally, you shouldn't need to modify your code to compile it for a Morello target. Morello (and CHERI) "just works" behind the scenes gently taking care of your spatial memory safety.
+
+When compiling more involved examples, you may see new compiler warnings, usually they are there for a good reason and may indicate some memory safety issue. This is one of those situations when you may need to change your code to let it compile without warnings for a Morello target. However, you will notice as well that after this change the same code builds nicely for a non-Morello target, and it even may work better because a bug may have been fixed.
+
+## Purecap and hybrid
+
+Now it's time to talk about purecap (pure capability) and hybrid Morello applications. From an application developer's point of view, in a purecap program all pointers automatically become capabilities. This means that all memory accesses will happen via (and be controlled by) some specific capability. On the other hand, in a hybrid application, pointers are addresses unless explicitly defined as capabilities. So, a hybrid program would work in the same way as a native (AArch64) program until it encounters a pointer that is explicitly defined as a capability.
```

Comment:
In general, we’ve made a transition to using the terms “CHERI C” and “CHERI C++” when talking about the language-level changes, and “pure cap” (etc) when talking about ABIs, representation of pointers in code, and so on. Hence the CHERI C/C++ programmer’s guide, which leaves open the possibility of different ABIs and code-generation models. I wonder if it could make sense to do the same here?

Reply:
I will clarify that I use these terms first of all in the context of pointer representation. The angle I want to use here is what happens when we compile (almost) the same code in different ways (making the compiler to invoke different rules and use different ABIs) to use the difference in behaviour to illustrate how capabilities work.